### PR TITLE
Add asm clobbers for PPC / PPC64

### PR DESCRIPTION
**Pull request by rrrogalski**
Marking this as a draft as I have a few questions :D

For the VMX ones, I get `error[E0599]: no variant or associated item named 'v19' found for enum 'PowerPCInlineAsmReg' in the current scope`. I'm fairly new to rust, and can't quite figure out what's triggering this :). Also, should these be included, when they're not available on every power CPU?

The second thing is this compiles without errors when I make the code specify PowerPC, but not when I make it for PowerPC64: `error[E0599]: no variant or associated item named 'PowerPC64' found for enum 'InlineAsmClobberAbi' in the current scope`. Some arches (like riscv) appear to use the same for 32bit and 64bit, while others have different clobbers for each. I'm pretty sure the registers correct for both 32 & 64 bit, but will PPC64 use this?

The third thing is r11 and r12. They're both volatile but have 'special' uses. 

```
r11       Volatile register used in calls by pointer and as an
          environment pointer for languages which require one
r12       Volatile register used for exception handling and glink code
```

I looked at other arches and their registers to see whether I should include these or not, but am unsure :). 

@Amanieu

### CODE REVIEW

File: compiler/rustc_target/src/asm/mod.rs [hunk 1]
```diff
@@ -885,6 +886,10 @@ impl InlineAsmClobberAbi {
                 "C" | "system" | "efiapi" => Ok(InlineAsmClobberAbi::LoongArch),
                 _ => Err(&["C", "system", "efiapi"]),
             },
+            InlineAsmArch::PowerPC => match name {
```
Comment (Amanieu) [inline compiler/rustc_target/src/asm/mod.rs:889]: You need to add `PowerPC64` here as well if both use the same calling convention. If the registers are different then you need a separate 32-bit and 64-bit `InlineAsmClobberAbi`.

**Comment by Amanieu** [comment]
> For the VMX ones, I get `error[E0599]: no variant or associated item named 'v19' found for enum 'PowerPCInlineAsmReg' in the current scope`. I'm fairly new to rust, and can't quite figure out what's triggering this :). Also, should these be included, when they're not available on every power CPU?

You need to define the VMX registers in compiler/rustc_target/src/asm/powerpc.rs. You'll need to create a `vreg` class for them.

> The second thing is this compiles without errors when I make the code specify PowerPC, but not when I make it for PowerPC64: `error[E0599]: no variant or associated item named 'PowerPC64' found for enum 'InlineAsmClobberAbi' in the current scope`. Some arches (like riscv) appear to use the same for 32bit and 64bit, while others have different clobbers for each. I'm pretty sure the registers correct for both 32 & 64 bit, but will PPC64 use this?

See code comments.

> The third thing is r11 and r12. They're both volatile but have 'special' uses.

All volatile registers need to be included. The point of `clobber_abi` is that it should include all registers that may be modified by a function call. This also needs to include the condition code registers.

**Comment by Richard-Rogalski** [comment]
Sorry, forgot pushing updates the PR. Still not done yet :)

File: compiler/rustc_target/src/asm/powerpc.rs [hunk 2]
```diff
@@ -47,6 +48,7 @@ impl PowerPCInlineAsmRegClass {
                 }
             }
             Self::freg => types! { _: F32, F64; },
+            Self::vreg => todo!(), //not quite sure :)
```
Comment (Amanieu) [inline compiler/rustc_target/src/asm/powerpc.rs:51]: I think this should be something like:
```rust
types! {
    vsx: I8, I16, I32, I64, F32, F64,
    altivec: VecI8(16), VecI16(8), VecI32(4), VecI64(2), VecF32(4), VecF64(2);
}
```
Altivec allows 128-bit vector types, and VSX allows using scalars in vector registers.

However this should be verified by writing asm tests that checks that LLVM is able to generate the correct code for these.

**Comment by Richard-Rogalski** [comment]
Should I make these tests in `tests/assembly`, or `tests/codegen`? and should I make a new file for these, or what `.rs` file would be most appropriate for these kinds of tests?

**Comment by Amanieu** [comment]
The tests should be added to `tests/assembly/asm/powerpc-types.rs`. You need to test every supported input type with the vreg class, and with an explicit vN register.

**Comment by Dylan-DPC** [comment]
@Richard-Rogalski any updates on this?

**Comment by Richard-Rogalski** [comment]
bah, yeah, I should get back to this. I got a little lost, I'll try again when i get the time

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Comment by Richard-Rogalski** [comment]
I've gotten better at rust in the time since I started this PR. If i am recalling correctly, everything is done except for the tests :p I'll take another crack at this starting today @ky438 .

**Pull request closed without merging**